### DISCUSSION TITLE
some examples of how to parse hostile JSON for the documentations?

### DISCUSSION THREAD

**Discussion by sfinktah** [nlohmann/json#2563] [Ideas]
> I added a `contains` function for JSON pointers, see [258fa79](https://github.com/nlohmann/json/commit/258fa798f15a0010d04eaa23010be8b53afd2f4d). Feedback is greatly appreciated! @sdaly2107 @brouer @nickaein

Great to hear, speaking of `contains`, I only found out about that one from looking at someone else's code (to whom I had recommended your library).  If it doesn't already exists, I believe the documentation could benefit from some examples of how to parse hostile JSON in situations where `throw`ing isn't an option, and you need to get as much valid data as possible.

I wrote this for 361, before `contains`, and they are not pretty to look at.  As improvements to the library occur, it would be nice to have an always up-to-date reference on safe parsing. (The option for handling invalid UTF-8 as binary was an unexpected bonus BTW).

FWIW, this is what can happen if one doesn't know a better way (if there **was** a better way -- now there is `.contains` but it would only shorted the macros) -- not pretty, but it never failed me.

```cpp
    // .contains() with a type check
    #define JSON_HAS(XJSON, XKEY, XTYPE) \
        (XJSON.find(XKEY) != XJSON.end() && XJSON.at(XKEY).is_##XTYPE())

    // everything above, plus a value comparison
    #define JSON_IS(XJSON, XKEY, XTYPE, XVALUE) \
        (XJSON.find(XKEY) != XJSON.end() && XJSON.at(XKEY).is_##XTYPE() && XJSON.at(XKEY) == XVALUE)


    do {
		auto j = json::parse(responseData);

		BREAK_UNLESS(JSON_HAS("_id", string));
		std::string _id = j["_id"];
		
		// ...
    

    if (auto iter = std::find_if(running.begin(), running.end(),
		[id](auto j) {		
		    return JSON_IS(j, "uid", number, id);
		}); iter != running.end()) { /* ... */

**Comment by nlohmann** [comment]
I am not completely sure what to make of this post, so here are some ideas:

- As for parsing without exceptions, see https://json.nlohmann.me/features/parsing/parse_exceptions/. There are also some examples.
- There is no proper overview page for parsing yet, see https://json.nlohmann.me/features/parsing/ - PRs welcome!
- There are examples in the API documentation, see https://json.nlohmann.me/api/basic_json/parse/ and https://json.nlohmann.me/api/basic_json/accept/
- Your scenario where you expect the JSON to have some pre-known structure could be a use-case for conversion into a user-defined struct or class, described in https://json.nlohmann.me/features/arbitrary_types/
